186.py

```diff
@@ -16,12 +16,10 @@ def reverse(start: int, end: int) -> None:
         while end < N:
             if s[end] == ' ':
                 reverse(start, end - 1)
-                end += 1
-                start = end
+                start = end + 1
             elif end == N - 1:
```

Comment:
The case of `end == N -  1` is not that special. It can be merged into the `s[end] == ' '` case.

Reply:
Never mind. I misread your logic. This implementation is good.